### DISCUSSION TITLE
How does "data-wp-effect" know when it has to be triggered so it doesn’t run more than needed?

### DISCUSSION THREAD

**Discussion by juanmaguitar** [WordPress/gutenberg#53019] [Interactivity API]
It seems this is handled by [Preact Signals](https://href.li/?https://preactjs.com/guide/v10/signals/), but the community could benefit from a deeper explanation.

**Comment by DAreRodz** [comment]
It's correct; let me expand a little bit.

Directives work as preact hooks, in the sense that they're functions called during a component rendering. And they can run other hooks, as hooks are _composable_. The `data-wp-effect` directive uses [`useSignalEffect`](https://preactjs.com/guide/v10/signals/#:~:text=When%20responding%20to%20signal%20changes%20within%20a%20component%2C%20use%20the%20hook%20variant%3A%20useSignalEffect(fn).) under the hood, which is the hook that `@preact/signals` offers to subscribe to signal changes, but only from those signals that are read during a component rendering. This means `data-wp-effect` will be triggered when the properties from the store **it uses** change, and only in that case.

I hope it's clearer now. 🙂

**Reply by luisherranz** [reply]
Exactly.

To add to David's excellent explanation: there are two _reactive_ objects in the Interactivity API: `context` (local) and `state` (global). Both of them use [deepsignal](https://github.com/luisherranz/deepsignal/), which takes an object and turns all its properties into Preact Signals.

This type of fine-grained reactivity works recursively, so if you call a function that reads from `context` or `state`, it will also trigger the reactivity. For example:

```js
const doubleFromState = (state) => state.myPlugin.counter * 2;
const doubleFromContext = (context) => context.myPlugin.counter * 2;

store({
  state: {
    myPlugin: {
      counter: 0,
    },
  },
  effects: {
    myPlugin: {
      logState: ({ state }) => {
        // Log every time state.myPlugin.counter changes.
        console.log(doubleFromState(state));
      },
      logContext: ({ context }) => {
        // Log every time context.myPlugin.counter changes.
        console.log(doubleFromContext(context));
      },
    },
  },
});
```
